### PR TITLE
feat(iOS): rewrite DRM Module

### LINKED ISSUE #4177: [BUG]: IOS DRM Custom License Request Not Working
### Version

"react-native-video": "^6.5.0"

### What platforms are you having the problem on?

iOS

### System Version

Mac mini M2

### On what device are you experiencing the issue?

Real device, Simulator

### Architecture

new architecture

### What happened?

Video error: {
  "error": {
    "localizedRecoverySuggestion": "",
    "code": 1718449215,
    "domain": "CoreMediaErrorDomain",
    "localizedDescription": "The operation couldn’t be completed. (CoreMediaErrorDomain error 1718449215.)",
    "localizedFailureReason": ""
  },
  "target": 3
}

### Reproduction

Step to reproduce this bug are: 

import React, { useRef } from 'react';
import { View, Platform, Alert } from 'react-native';
import Video from 'react-native-video';

const VideoPlayer = () => {
  const isIos = Platform.OS === 'ios';
  const videoRef = useRef(null);

  
  return (
    <View style={{ flex: 1 }}>
      <Video
        ref={videoRef}
        source={{
          uri: isIos
            ? 'https://d2a1rntxelqk4o.cloudfront.net/transcoded/xxxxxxxxxxx/video.m3u8'
            : 'https://d384padtbeqfgy.cloudfront.net/transcoded/xxxxxxxxxxx/video.mpd',
          type: isIos ? 'm3u8' : 'mpd',
          drm: {
                type: 'fairplay',
                certificateUrl: `https://app.tpstreams.com/static/fairplay.cer`,
                licenseServer: 'https://app.tpstreams.com/api/v1/xxxxxx/assets/xxxxxxxxxxx/drm_license/?access_token=xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx&drm_type=fairplay',
                base64Certificate: true,
                getLicense: async (spcBase64, contentId, licenseUrl) => {
                  return await fetch('https://app.tpstreams.com/api/v1/xxxxxx/assets/xxxxxxxxxxx/drm_license/?access_token=xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx&drm_type=fairplay', {
                    method: 'POST',
                    headers: {
                      'Content-Type': 'application/x-www-form-urlencoded',
                    },                     
                  }).then(response => response.text()).then((response) => {
                    console.log(response);
                    return response;
                  }).catch((error) => {
                    console.error('Error', error);
                  });
                }
          
      }

        }}
        style={{ flex: 1 }}
        controls
        onLoad={(data) => console.log('Video loaded:', data)}
        onReadyForDisplay={() => console.log('Video ready for display')}
        onError={(error) => {
          console.log('Video error:', JSON.stringify(error, null, 2));
        
        }}
      />
    </View>
  );
};

export default VideoPlayer;

### PR DESCRIPTION
**Pull request by KrzysztofMoch**
## Summary
I have re-implemented DRM module for Apple devices. New Implementation use `AVContentKeySession` instead of resource loader.

fixes #4177 #3675 #3467

### Motivation
Motivation is to fix issue with DRM Airplay and add visionOS support (in old implementation it was not possible)

## Test plan
- I have test app with Fairplay stream (which unfortunately I cannot share)
- Also to play DRM content you need to have real device, simulator will not work

### Tested on
- iOS is tested by me
- tvOS is not tested and I don't have real device to do it
- visionOS is not tested and I don't have real device to do it

### CODE REVIEW

File: ios/Video/Features/DRMManager+AVContentKeySessionDelegate.swift
Comment (freeboub) [inline ios/Video/Features/DRMManager+AVContentKeySessionDelegate.swift:2]: Is there a reason to have + in file name ? (I don't really like that, but that's only my point of view ... )

File: ios/Video/Features/RCTVideoErrorHandling.swift
Comment (freeboub) [inline ios/Video/Features/RCTVideoErrorHandling.swift:77]: This is a good rework, but I don't like these strings.
For app developement, it would be better to have some kind of error code. When you need to internationnalize your app working with such descriptions is not really friendly. (moreover they are not documented).
I don't give you more clues because it may need breaking changes, this is a good rework anyway and you don't break the rnv api for this point...

File: ios/Video/Features/DRMManager+AVContentKeySessionDelegate.swift
Comment (KrzysztofMoch) [reply ios/Video/Features/DRMManager+AVContentKeySessionDelegate.swift:2]: This is for files with extension. It's ClassName+Extension - I see this in other code spaces and I like it bcs I see that this file is extension for DRMManager (in this case) - if you don't like it I guess I can change it

File: ios/Video/Features/RCTVideoErrorHandling.swift
Comment (KrzysztofMoch) [reply ios/Video/Features/RCTVideoErrorHandling.swift:77]: How about making "Error Rework" that will standardise errors and will give better DevX - I can do it in separate PR WDYT ?

Comment (freeboub) [reply ios/Video/Features/RCTVideoErrorHandling.swift:77]: Yep, I think it should be done in a separated PR!

**Comment by KrzysztofMoch** [comment]
@freeboub should I add someone else to reviewers ?

**Comment by freeboub** [comment]
Maybe @paul-rinaldi or @paulrinaldi would like to test this version ? I can review but not test on my side ...

File: ios/Video/Features/DRMManager+AVContentKeySessionDelegate.swift
Comment (freeboub) [inline ios/Video/Features/DRMManager+AVContentKeySessionDelegate.swift]: DebugLog would be better ?

File: ios/Video/Features/RCTResourceLoaderDelegate.swift
Comment (freeboub) [inline ios/Video/Features/RCTResourceLoaderDelegate.swift:141]: looks like you are no more checking drmType value ?

File: ios/Video/Features/RCTVideoDRM.swift
Comment (freeboub) [inline ios/Video/Features/RCTVideoDRM.swift:54]: I don't find this urlEncoded also ?

File: ios/Video/RCTVideo.swift
Comment (freeboub) [inline ios/Video/RCTVideo.swift:454]: not sure what is the usage of this `localSourceEncryptionKeyScheme` is it still supported ?!

Comment (freeboub) [inline ios/Video/RCTVideo.swift:465]: localSourceEncryptionKeyScheme should be moved inside drm prop no ?

**Comment by freeboub** [comment]
I added some small comments, but it looks like good job 👍 
If you are looking for other testers, you can try to ping last issue openers which are linked to DRM on ios. they may have a look to the PR or test it !

**Comment by paul-rinaldi** [comment]
> I added some small comments, but it looks like good job 👍 
> If you are looking for other testers, you can try to ping last issue openers which are linked to DRM on ios. they may have a look to the PR or test it !

I don't have a tv with iOS on it or vision OS. I have physical iPhones and physical iPads I could test with if you would like me to but it sounds like the PR creator already did that.

Sorry I could not be more helpful.

**Comment by freeboub** [comment]
> > I added some small comments, but it looks like good job 👍 
> > If you are looking for other testers, you can try to ping last issue openers which are linked to DRM on ios. they may have a look to the PR or test it !
> 
> I don't have a tv with iOS on it or vision OS. I have physical iPhones and physical iPads I could test with if you would like me to but it sounds like the PR creator already did that.
> 
> Sorry I could not be more helpful.

My proposal was linked to your message on discord in fact !

**Comment by roo12312** [comment]
> I added some small comments, but it looks like good job 👍 
> If you are looking for other testers, you can try to ping last issue openers which are linked to DRM on ios. they may have a look to the PR or test it !

I have tested it and it works fine also it solved buffering issues in drm videos.

File: ios/Video/RCTVideo.swift
Comment (freeboub) [reply ios/Video/RCTVideo.swift:465]: This comment is not mandatory to merge the PR !

File: src/Video.tsx
Comment (freeboub) [inline src/Video.tsx:193]: Thanks @KrzysztofMoch , remember to update the doc !

File: ios/Video/Features/RCTResourceLoaderDelegate.swift
Comment (KrzysztofMoch) [reply ios/Video/Features/RCTResourceLoaderDelegate.swift:141]: Good catch! fixed

File: ios/Video/RCTVideo.swift
Comment (KrzysztofMoch) [reply ios/Video/RCTVideo.swift:465]: I moved it but without braking change - deprecated old one
Forgot to update docs - I will do it soon

File: src/Video.tsx
Comment (KrzysztofMoch) [reply src/Video.tsx:193]: Yup! will do it

File: ios/Video/Features/RCTVideoDRM.swift
Comment (KrzysztofMoch) [reply ios/Video/Features/RCTVideoDRM.swift:54]: I skiped this because I was not sure what was it doing? it adding spc as base64 to url as query param ?

Comment (freeboub) [reply ios/Video/Features/RCTVideoDRM.swift:54]: Yes it looks like ... 
but as far as I remember the SPC shall be provided in the body 🤔 
How did you test ?
Did you declare getLicenseUrl in JS ?
because I think the package have 2 different ways of working one with getLicense but I remember it is also possible to handle internally only...
Deep doc review necessary 👀

Comment (KrzysztofMoch) [reply ios/Video/Features/RCTVideoDRM.swift:54]: In my test app I have getLicense. From what I found in internet SPC is usually in body. Now "internally" I am adding spc in body

Comment (freeboub) [reply ios/Video/Features/RCTVideoDRM.swift:54]: Ok, let see 🤷

**Review by freeboub — APPROVED** [review]

**Comment by freeboub** [comment]
Looks good to me. Please build a release after merging this PR :)

**Comment by KrzysztofMoch** [comment]
Sure I will give this PR to tests for our client and after it will do it! Thanks!

**Comment by roo12312** [comment]
Waiting for the merge

File: ios/Video/RCTVideo.swift
Comment (freeboub) [inline ios/Video/RCTVideo.swift]: The logic change here ?

File: docs/pages/component/props.mdx
Comment (freeboub) [inline docs/pages/component/props.mdx]: can you pleas just copy this section in the DRM documentation ?

File: ios/Video/RCTVideo.swift
Comment (KrzysztofMoch) [reply ios/Video/RCTVideo.swift]: Whops, will fix rebase!

Comment (KrzysztofMoch) [reply ios/Video/RCTVideo.swift]: On master we have `source.textTracks.isEmpty != true` - this is expected ?

Comment (freeboub) [reply ios/Video/RCTVideo.swift]: I think you are right in fact, give me 10 min to double check

Comment (freeboub) [reply ios/Video/RCTVideo.swift]: @KrzysztofMoch OK, You are fixing a regression I introduced ...

Comment (KrzysztofMoch) [reply ios/Video/RCTVideo.swift]: it should be fine now

File: docs/pages/component/props.mdx
Comment (KrzysztofMoch) [reply docs/pages/component/props.mdx]: Done

Comment (freeboub) [reply docs/pages/component/props.mdx]: 👍

**Review by freeboub — APPROVED** [review]

File: ios/Video/RCTVideo.swift
Comment (KrzysztofMoch) [inline ios/Video/RCTVideo.swift]: @freeboub `source.drm` is never `nil` is structure with `nils` even if we don't pass anything are you okay with checking `drm.type` ?

Comment (freeboub) [reply ios/Video/RCTVideo.swift]: I am not sure to understand, you want to change 
    let drm: DRMParams?
by 
    let drm: DRMParams

And check only type ?
if that's the case, I agree !

Something like this patch:
```diff
diff --git a/ios/Video/DataStructures/VideoSource.swift b/ios/Video/DataStructures/VideoSource.swift
index e672929f..45e5b2d3 100644
--- a/ios/Video/DataStructures/VideoSource.swift
+++ b/ios/Video/DataStructures/VideoSource.swift
@@ -10,7 +10,7 @@ struct VideoSource {
     let cropEnd: Int64?
     let customMetadata: CustomMetadata?
     /* DRM */
-    let drm: DRMParams?
+    let drm: DRMParams
     var textTracks: [TextTrack] = []
 
     let json: NSDictionary?
@@ -28,7 +28,7 @@ struct VideoSource {
             self.cropStart = nil
             self.cropEnd = nil
             self.customMetadata = nil
-            self.drm = nil
+            self.drm = DRMParams(nil)
             return
         }
         self.json = json
diff --git a/ios/Video/RCTVideo.swift b/ios/Video/RCTVideo.swift
index 908e3eb8..64130675 100644
--- a/ios/Video/RCTVideo.swift
+++ b/ios/Video/RCTVideo.swift
@@ -421,7 +421,7 @@ class RCTVideo: UIView, RCTVideoPlayerViewControllerDelegate, RCTPlayerObserverH
                 "type": _source?.type ?? NSNull(),
                 "isNetwork": NSNumber(value: _source?.isNetwork ?? false),
             ],
-            "drm": source.drm?.json ?? NSNull(),
+            "drm": source.drm.json ?? NSNull(),
             "target": reactTag as Any,
         ])
 ```

Comment (KrzysztofMoch) [reply ios/Video/RCTVideo.swift]: I meant:
in current impl source.drm is never null it will be always struct with some values (if drm is not setted but user all values will be just null). But now as I think I can aloso check `if source.drm.json != nil` bcs if json is null it means that users notting passed to it

File: ios/Video/Features/RCTVideoErrorHandling.swift
Comment (agastiazoro) [inline ios/Video/Features/RCTVideoErrorHandling.swift]: @KrzysztofMoch there is a typo in "content"

**Comment by maksymhcode-care** [comment]
Is this ready to be merged?

File: ios/Video/RCTVideo.swift
Comment (freeboub) [inline ios/Video/RCTVideo.swift]: @KrzysztofMoch I am reviewing use of String! and it is a bad practice and can lead to some issues, see: https://github.com/TheWidlarzGroup/react-native-video/issues/4181

It is just hide the warning, so no need to test the value ...
I think it would be better to change like I propose (not sure of the consequences inside _drmManager)

```suggestion
    func setLicenseResult(_ license: String?, _ licenseUrl: String?) {
```

File: ios/Video/RCTVideo.swift
Comment (freeboub) [inline ios/Video/RCTVideo.swift]: Same here:
```suggestion
    func setLicenseResultError(_ error: String?, _ licenseUrl: String?) {
```

File: ios/Video/Features/RCTVideoErrorHandling.swift
Comment (agastiazoro) [inline ios/Video/Features/RCTVideoErrorHandling.swift]: @KrzysztofMoch there is a typo in "unsupported"

File: ios/Video/RCTVideo.swift
Comment (KrzysztofMoch) [reply ios/Video/RCTVideo.swift]: I changed it to just `String` - I don't like idea of making it `String?` as it gives idea that those params are not required when they are

File: ios/Video/RCTVideo.swift
Comment (KrzysztofMoch) [reply ios/Video/RCTVideo.swift]: https://github.com/TheWidlarzGroup/react-native-video/pull/4136#discussion_r1768780829

File: ios/Video/RCTVideo.swift
Comment (freeboub) [reply ios/Video/RCTVideo.swift]: But the app can call it with undefined parameter, in that case we should be robust. Am I wrong ?
This is not a mandatory change!

Comment (KrzysztofMoch) [reply ios/Video/RCTVideo.swift]: Alright then we should made changes here (but lets do it in other PR):
https://github.com/TheWidlarzGroup/react-native-video/blob/c96f7d41f38499baefde2b7fa7cf10441463438e/ios/Video/RCTVideoManager.swift#L40-L45

**Pull request merged**
